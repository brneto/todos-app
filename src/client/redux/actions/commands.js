import { createActions } from 'redux-actions';
import { metaFilterListCreator } from './utils';

// Pure Command actions
// Naming Convention: <imperative verb><subject>
export const { addTodoToList, removeTodoFromList } =
  createActions({
    ADD_TODO_TO_LIST: metaFilterListCreator,
    REMOVE_TODO_FROM_LIST: metaFilterListCreator,
  });
