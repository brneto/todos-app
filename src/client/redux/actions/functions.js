import { identity } from '../reducers/util';

// actionCreator :: type -> (data, filter | undefined) -> {
//   type: 'COMMAND' | 'DOCUMENT' | 'EVENT' | 'SIDE_EFFECT',
//   payload: data,
//   meta: { type: type | [, filter: filter] }
// }

export const createActionFunction = type => [
  identity,
  (payload, filter) => filter ? ({ type, filter }) : ({ type })
];
