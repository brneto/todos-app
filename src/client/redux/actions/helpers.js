
import { identity } from 'ramda';

// actionCreator :: type -> (data, filter | undefined) -> {
//   type: 'ACTION_TYPE',
//   payload: data,
//   meta: { type: type | [, filter: filter] }
// }
const
  createActionFunction = type =>
    [identity, (payload, filter) => filter ? ({ type, filter }) : ({ type })];

export { createActionFunction };
