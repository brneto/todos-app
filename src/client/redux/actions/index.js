import { createActions } from 'redux-actions';

export const {
  fetchTodos,
  addTodo,
  toggleTodo,

  setToggleFetching,

  setFetchTodos,
  setAddTodo,
  setToggleTodo,
} = createActions(
  {
    SET_TOGGLE_FETCHING: [
      () => null,
      (payload, meta) => ({ filter: payload instanceof Error ? meta : payload })
    ],
    SET_FETCH_TODOS: [
      payload => payload,
      (todos, filter) => ({ filter })
    ],
  },
  'SET_ADD_TODO',
  'SET_TOGGLE_TODO',
  'FETCH_TODOS',
  'ADD_TODO',
  'TOGGLE_TODO',
);
