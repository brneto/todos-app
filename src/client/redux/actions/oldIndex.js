import { createActions } from 'redux-actions';
import { identity } from 'ramda';

// TODO: Breaks this file into event types messages
const filterHandler = [identity, (_, filter) => ({ filter })];
// setFetchedTodos :: (a, b) -> {
//   type: 'SET_FETCHED_TODOS',
//   payload: a,
//   meta: { filter: b }
// }
export const {
  // Event messages
  // Naming Convention: ?
  setToggleFetching, // toggleFetching

  // Command(Modifier/Mutator) messages
  // Naming Convention: <imperative verb><subject>
  setToggledTodoAdd, // addTodoToList(['completed'|'active'])
  setToggledTodoRemove, // removeTodoFromList(['completed'|'active'])
} = createActions(
  {
    SET_TOGGLED_TODO_ADD: filterHandler,
    SET_TOGGLED_TODO_REMOVE: filterHandler,
  },
  'SET_TOGGLE_FETCHING',
);
