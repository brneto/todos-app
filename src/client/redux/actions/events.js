import { createActions } from 'redux-actions';
import { createActionFunction } from './functions';

const messageType = 'EVENT';

// Event actions
// Naming Convention:
// - Instant event: <present-tense verb>
// - Progressive event start: <present-continuous verb>[<subject>]
// - Progressive event end: <past-tense verb>[<subject>]
export const {
  fetchingTodos, fetchedTodos
} = createActions({
  'FETCHING_TODOS': createActionFunction(messageType),
  'FETCHED_TODOS': createActionFunction(messageType),
});
