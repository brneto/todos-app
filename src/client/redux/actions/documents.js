import { createActions } from 'redux-actions';
import { createActionFunction } from './helpers';

const messageType = 'DOCUMENT';

// Document actions
// Naming Convention: <subject><past-tense verb>
export const {
  todosFetched, todoAdded, todoToggled
} = createActions({
  TODOS_FETCHED: createActionFunction(messageType),
  TODO_ADDED: createActionFunction(messageType),
  TODO_TOGGLED: createActionFunction(messageType),
});
