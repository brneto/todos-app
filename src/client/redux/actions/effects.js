import { createActions } from 'redux-actions';
import { createActionFunction } from './helpers';

const messageType = 'SIDE_EFFECT';

// Side Effect Command actions
// Naming Convention: <imperative verb><subject>
export const {
  fetchTodos, addTodo, toggleTodo
} = createActions({
  'FETCH_TODOS': createActionFunction(messageType),
  'ADD_TODO': createActionFunction(messageType),
  'TOGGLE_TODO': createActionFunction(messageType),
});
