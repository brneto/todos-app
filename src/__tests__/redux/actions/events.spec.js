import { events } from '../../../client/redux/actions';

describe('Event action messages', () => {
  const filter = 'filterName';

  test('fetchingTodos event', () => {
    expect(events.fetchingTodos(filter)).toMatchSnapshot();
  });

  test('fetchedTodos event', () => {
    expect(events.fetchedTodos(filter)).toMatchSnapshot();
  });
});

