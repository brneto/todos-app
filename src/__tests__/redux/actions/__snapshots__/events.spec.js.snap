// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event action messages fetchedTodos event 1`] = `
Object {
  "meta": Object {
    "type": "EVENT",
  },
  "payload": "filterName",
  "type": "FETCHED_TODOS",
}
`;

exports[`Event action messages fetchingTodos event 1`] = `
Object {
  "meta": Object {
    "type": "EVENT",
  },
  "payload": "filterName",
  "type": "FETCHING_TODOS",
}
`;
